Human-readable encoding 1: serialization and README

apoelstra opened the pull request:
This PR introduces the human-readable serialization of Simplicity programs. Right now it will serialize disconnect nodes as having only one child (and the next PR will parse such programs). Later, when we introduce typed holes, we will correct this so that disconnect nodes always have two children (the rightmost one being a named hole). But for now disconnect is more-or-less unusable. To avoid too much complexity at once I think this is a reasonable order of operations.

It also slightly changes the syntax from the last PR to allow `'` in symbol names. This is so I can do stuff like `x' := pair x unit` or whatever, where I use the prime symbol to indicate a slightly tweaked version of a different expression.

apoelstra commented:
I have a crappy CLI tool that I've been mostly interacting with this code through. If you'd like I can clean that up a little and add it to this PR, so you can feed it hex-encoded Simplicity programs and get the serialization.

Could also finally switch to base64..

sanket1729 commented:
I started reviewing this PR. Only read the README so far. Will play around with this over the next couple days. 


Any CLI tool would be helpful. it doesn't have to be cleaned up. I can navigate through the CLI code to figure out stuff

apoelstra commented:
@sanket1729 you may also be interested in the full branch (which I keep reasonably up to date) https://github.com/BlockstreamResearch/rust-simplicity/tree/2023-06--simplang which includes the CLI tool and also parsing logic.

apoelstra commented:
Updated README and also removed the `name_map` from `Program` because I don't use it yet and it was complicating some later code. Will restore it in a later PR.

uncomputable commented on src/human_encoding/mod.rs:
```diff
@@ -0,0 +1,55 @@
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Program<J: Jet> {
+    root: Arc<NamedCommitNode<J>>,
+}
```
cf2ea20bbbf0297347b132d7c1b382e88f8997e4: We should think about how we handle programs of other node types. So far, we use `&Node` or `Arc<Node>` in methods like `BitMachine::for_program()`.

apoelstra replied:
Well, right now we take a `RedeemNode` in the bit machine because that's the only node type that is guaranteed to have witness values, complete types, and disconnected expressions. If we add another such node type we'll need to add a trait describing this.

apoelstra replied:
Nothing in my PR stack involves a `NamedRedeemNode` and I haven't found a use for one ... I think, if you want to execute a program, you should just convert it to `RedeemNode` beforehand.

uncomputable commented on src/human_encoding/mod.rs:
```diff
@@ -52,4 +54,96 @@ impl<J: Jet> Program<J> {
+            ret += "\n";
+        }
+        if !const_lines.is_empty() {
+            // FIXME detect scribes
```
0fe9d92823b1aaddd39ca0ae83374db1884205f0: [I have some code that detects scribe.](https://github.com/uncomputable/hal-simplicity/blob/master/src/compress.rs) I can rebase it and make a PR later if you want.

apoelstra replied:
That'd be awesome, though I think we need to decide what to do about const word jets and whether a "scribe" includes them.

apoelstra replied:
I'd also hold off on this until the parser is PR'd.

uncomputable commented on src/human_encoding/README.md:
```diff
@@ -0,0 +1,189 @@
+Comments are started by `--` and end at the next newline. This is the only aspect
+in which whitespace is significant.
+
+Simplicity expressions are a series of **definitions** of the form
```
aed42d0045f0403f05e263c737bf6dc3c4153d9a: The word "expression" has a different meaning here than in the rest of the document. A (semantic) Simplicity expression is defined by a series of (syntactic) definitions.

Maybe "A Simplicity file consists of a series of DEFINITIONS of the form..."

apoelstra replied:
Yes, "Simplicity expression" has a different meaning than "expression". I am careful about this but I agree that it's confusing and potentially misleading.

uncomputable commented on src/human_encoding/README.md:
```diff
@@ -0,0 +1,189 @@
+In general, we do not need to obtain a single expression. It is permissible to
+encode a "DAG forest".
+
+The name `main` is special for several reasons:
+* An expression named `name` implicitly has the type ascription `1 -> 1`. That is, it must always be a program.
+* To generate a commitment-time program from an expression, it must be called `main`.
+* Type ascriptions for `main` and its children are enforced **after** type inference has completed, so they act as sanity checks but cannot change the output of type inference. For other expressions, type ascriptions are enforced **before** and may guide inference.
```
aed42d0045f0403f05e263c737bf6dc3c4153d9a: We should define what a Simplicity program is, and also what a (semantic) Simplicity expression is.

apoelstra replied:
Yes, but probably not in the README for the human-readable encoding.

uncomputable commented on src/human_encoding/README.md:
```diff
@@ -0,0 +1,189 @@
+ignored by the parser.
+
+A HOLE is the literal `?` followed by a NAME. It indicates an expression that has
+yet to be defined. Holes have a different namespace than other names.
```
```suggestion
yet to be defined. Holes have a different namespace than names / NAMES.
```

apoelstra commented on src/human_encoding/serialize.rs:
```diff
@@ -0,0 +1,37 @@
+        } else {
+            f.write_str("0b")?;
+            self.0
+                .do_each_bit(|b| f.write_str(if b { "1" } else { "0" }).expect("FIXME"));
```
I think we should actually just use the `PreOrderIter` which is made for exactly this and is a little more Rustic. (In Rust it's usually clearer to use external (for-loop-based) rather than internal (closure-based) iteration.) This FIXME predates `Value` being `DagLike`.

apoelstra replied:
BTW @uncomputable for future reference, to upload patches to PRs, you can literally just push (or force-push) to the branch on my repo. Because I opened a PR here, Github has made that branch writeable to maintainers of this repo.

uncomputable commented on simpcli/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+    eprintln!("For commands which take an optional expression, the default value is \"main\".");
+    eprintln!();
+    eprintln!("Run `{} help` to display this message.", process_name);
+    Err("invalid usage".into())
```
2459f566bfea051bec2225f82bbdf68a69abfcd1: We should not return an error on `simpcli help`.

uncomputable reviewed (commented):
2459f566bfea051bec2225f82bbdf68a69abfcd1: Looks good. There are some comments. The biggest question is what "expression" should mean.

apoelstra commented:
Force pushed to

* fix FIXME in the serialization logic for constant words
* update README in a couple places, most notably to change "definition" to "named expression"

apoelstra commented:
* Got rid of the confusingly-named `Program` type in favor of `Forest`.
* Added a CI job which just runs `cargo test` on the simpcli crate.

No other changes.

uncomputable reviewed (commented):
851928f4f61082f3773011dbabb8b6279568b23a: Looks good. I think we should split "named expressions" into "definitions" and "type bounds" / "declarations". We should also emphasize that there are three namespaces: one for expressions, one for holes and one for types.

apoelstra commented:
Changed `::` to `:` and updated README again.

We have another TODO to change `#` to `#{...}` but I think we can do that in a followup PR. Up to you.
